src/Analyzer/Passes/QueryAnalysisPass.cpp

```diff
@@ -2857,23 +2919,29 @@ QueryTreeNodePtr QueryAnalyzer::tryResolveIdentifierFromAliases(const Identifier
 {
     const auto & identifier_bind_part = identifier_lookup.identifier.front();
 
-    auto get_alias_name_to_node_map = [&]() -> std::unordered_map<std::string, QueryTreeNodePtr> &
-    {
-        if (identifier_lookup.isExpressionLookup())
-            return *scope.alias_name_to_expression_node;
-        else if (identifier_lookup.isFunctionLookup())
-            return scope.alias_name_to_lambda_node;
+    // auto get_alias_name_to_node_map = [&]() -> std::unordered_map<std::string, QueryTreeNodePtr> &
+    // {
+    //     if (identifier_lookup.isExpressionLookup())
+    //         return *scope.alias_name_to_expression_node;
+    //     else if (identifier_lookup.isFunctionLookup())
+    //         return scope.alias_name_to_lambda_node;
 
-        return scope.alias_name_to_table_expression_node;
-    };
+    //     return scope.alias_name_to_table_expression_node;
+    // };
+
+    // auto & alias_name_to_node_map = get_alias_name_to_node_map();
+    // auto it = alias_name_to_node_map.find(identifier_bind_part);
 
-    auto & alias_name_to_node_map = get_alias_name_to_node_map();
-    auto it = alias_name_to_node_map.find(identifier_bind_part);
+    // if (it == alias_name_to_node_map.end())
+    //     return {};
 
-    if (it == alias_name_to_node_map.end())
+    auto it = scope.aliases.find(identifier_lookup, ScopeAliases::FindOption::FIRST_NAME);
+    if (it == nullptr)
         return {};
 
-    if (!it->second)
+    QueryTreeNodePtr & alias_node = *it;
+
+    if (!alias_node)
         throw Exception(ErrorCodes::LOGICAL_ERROR,
             "Node with alias {} is not valid. In scope {}",
             identifier_bind_part,
```

Comment:
I think it's better to move this check inside of `ScopeAliases` and make `ScopeAliases::find` return `QueryTreeNodePtr` instead of `QueryTreeNodePtr *`. This line checks that the alias map is filled with correct data.

Reply:
I don't know how to do it properly:
* There is a place where it's possible to not find the alias: https://github.com/ClickHouse/ClickHouse/blob/51afec49107864e97eb36f9e5760efd1e11bfea8/src/Analyzer/Passes/QueryAnalysisPass.cpp#L4198-L4204
* We can replace alias node to resolved one later https://github.com/ClickHouse/ClickHouse/blob/51afec49107864e97eb36f9e5760efd1e11bfea8/src/Analyzer/Passes/QueryAnalysisPass.cpp#L2987

So I can't return `QueryTreeNodePtr` because it won't be possible to update the alias map, and I can't use `QueryTreeNodePtr &` because it's ok to not find an alias, and I can't use `std::optional<QueryTreeNodePtr &>` because this is not supported.